train_gpt2.cu

```diff
@@ -1898,6 +2024,41 @@ void gpt2_backward(GPT2 *model) {
     encoder_backward(grads.wte, grads.wpe, dresidual, model->inputs, B, T, C);
 }
 
+// Compute a mean of a single CPU value across all GPU processes. No-op when multi-GPU is disabled.
+float multi_gpu_cpu_float_mean(float value, const MultiGpuConfig* multi_gpu_config) {
+#ifdef MULTI_GPU
+    // MPI doesn't support all reduce with mean, so we sum up, then divide.
+    float result;
+    mpiCheck(MPI_Allreduce(&value, &result, 1, MPI_FLOAT, MPI_SUM, MPI_COMM_WORLD));
+    return result / multi_gpu_config->num_processes;
+#else
+    return value;
+#endif
+}
+
+// Averages out the loss and gradients across all GPUs. No-op when multi-GPU is disabled.
+void gpt2_mutli_gpu_accumulate(GPT2* model, MultiGpuConfig* multi_gpu_config) {
+    // Average all losses.
+    model->accumulated_mean_loss = multi_gpu_cpu_float_mean(model->mean_loss, multi_gpu_config);
+#ifdef MULTI_GPU
+    // Average all gradients.
+    char* grads_memory_iterator = (char*)model->grads_memory;
+    for (int i = 0; i < NUM_PARAMETER_TENSORS; ++i) {
```

Comment:
ugh really sad to have this loop here :( with @ngc92 changed this will be just two calls. but ideally it would be 1.

Reply:
I agree it is ugly. Looking forward to the rearrangements.

I think we would need at least 2 calls, as long as we have both 16 and 32 bit gradients.